### DISCUSSION TITLE
[KC17] [Logging] Running Keycloak as a container: how to send logs to logstash ?

### DISCUSSION THREAD

**Discussion by semangard** [keycloak/keycloak#10884] [Archive]
Hello,

We are running KeyCloak as a containers' cluster for a while on Docker Swarm.
For some reasons we need to send our logs to ELK through logstash.

With previous version this was possible with the following CLI :
```

# ##############################################################################
# Adjust default KeyCloak logging settings to send then to Logstash > ELK
# WARNING : commands CANNOT be ran by a batch (due to the if test and try-catch)
# http://docs.wildfly.org/23/Admin_Guide.html#Logging
# https://www.wildfly.org/news/2020/01/04/Custom-Logging-Filter/
# https://docs.wildfly.org/20/wildscribe/core-service/module-loading/index.html
# ##############################################################################

embed-server --server-config=standalone-ha.xml --std-out=echo

echo "--------------------------------------------------------------------------"

# ##############################################################################
# LOGSTASH handler
# ##############################################################################
try
    if (outcome == success) of /core-service=module-loading:module-info(name=ch.hcuge.logging)
        echo "Logging module already added => removing it..."
        module remove --name=ch.hcuge.logging
    end-if
catch
end-try
try
    echo "Adding LOGGING module..."
    module add --name=ch.hcuge.logging --resources=/opt/jboss/keycloak/providers/hug-keycloak.jar --dependencies=org.jboss.modules,java.logging,org.jboss.logging
catch
    echo "ERROR while adding logging module"
end-try
try
    echo "Adding custom filter..."
    /subsystem=logging/filter=customFilter:add(module=ch.hcuge.logging, class=ch.hcuge.logging.KeycloakMDCFilter)
catch
end-try
try
    echo "Adding JSON formatter..."
    /subsystem=logging/json-formatter=JSON-FORMATTER/:add(zone-id=GMT+1)
catch
end-try

# JUST for DEBUGGING JSON formatter
# try
#    echo "Updating console formatter..."
#    /subsystem=logging/console-handler=CONSOLE/:write-attribute(name=named-formatter,value=JSON-FORMATTER)
# catch
# end-try

try
    echo "Adding a socket binding..."
    /socket-binding-group=standard-sockets/remote-destination-outbound-socket-binding=log-server:add(host="${env.LOGSTASH_DESTINATION:logstash}", port="${env.LOGSTASH_PORT:9689}")
catch
end-try
try
    echo "Adding custon filer to socket handler..."
    /subsystem=logging/socket-handler=log-server-handler:add(filter-spec=customFilter, outbound-socket-binding-ref=log-server, block-on-reconnect=true, named-formatter=JSON-FORMATTER,protocol=TCP, level="${env.LOGSTASH_LOG_LEVEL:DEBUG}")
    # /subsystem=logging/async-handler=async-log-server-handler:add(level=ALL, queue-length=${env.LOGSTASH_QUEUE:1000}, overflow-action=BLOCK,subhandlers=[log-server-handler])
catch
end-try
try
    echo "Adding the log handler..."
    /subsystem=logging/root-logger=ROOT:add-handler(name=log-server-handler)
catch
end-try

# ##############################################################################
# CONSOLE handler
# ##############################################################################
# add yyyy-MM-dd GMT+1 to each CONSOLE log
try
    echo "Changing CONSOLE pattern..."
    /subsystem=logging/pattern-formatter=COLOR-PATTERN:write-attribute(name="pattern", value="%K{level}%z{GMT+1}%d{yyyy-MM-dd HH:mm:ss,SSS XXX} %-5p [%c] (%t) %s%e%n")
catch
end-try

echo "--------------------------------------------------------------------------"

stop-embedded-server

```

And thus we just had to set the following env vars on our container spec:
```
 LOGSTASH_DESTINATION: ...
 LOGSTASH_PORT: ...
 LOGSTASH_QUEUE: ...
```

==> How to do something similar with KeyCloak 17 (on Quarkus) runned as a container  (also JSON logs...) ?

Thanks and regards

**Comment by tse00** [comment]
Hello,

we are currently also working on the setup for Logstash on Keycloak 17. Unfortunately after I made all steps needed, I still wasn't able to see any logs from Keycloak 17 on Logstash/Kibana. It seems like quarkus/keycloak don't recognize the gelf plugin since it doens't appear on the logs under "_Installed features_".


![Bildschirmfoto 2022-03-24 um 17 07 13](https://user-images.githubusercontent.com/9500398/159960580-7107dc9b-7984-411c-9920-e5f858e34f82.png)


Semangard I hope this help you. With the help of the documentation and some information found on the internet, we got this setup:

- copied quarkus-logging-gelf-2.7.0.jar to `/providers` and to `/opt/keycloak/lib/lib/main/` (not sure if the last path is necessary)

- created `conf/quarkus.properties` file and add the quarkus-gelf parameters:

  ```
  quarkus.log.handler.gelf.enabled=true
  quarkus.log.handler.gelf.host=...
  quarkus.log.handler.gelf.port=...
  quarkus.log.handler.gelf.facility=...
  ```
  More info: https://quarkus.io/guides/centralized-log-management

- Our Dockerfile looks something like this:

  ```
  FROM quay.io/keycloak/keycloak:17.0.0 as builder
  
  ENV KC_DB=postgres
  ENV KC_FEATURES=quarkus-logging-gelf
  
  # QUARKUS GELF
  COPY ./docker/conf/quarkus.properties /opt/keycloak/conf/quarkus.properties
  COPY ./docker/lib/lib/main/quarkus-logging-gelf-2.7.0.Final.jar /opt/keycloak/lib/lib/main/quarkus-logging-gelf-2.7.0.Final.jar
  COPY ./docker/providers/quarkus-logging-gelf-2.7.0.Final.jar /opt/keycloak/providers/quarkus-logging-gelf-2.7.0.Final.jar
  
  RUN /opt/keycloak/bin/kc.sh build
  
  FROM quay.io/keycloak/keycloak:17.0.0
  
  COPY --from=builder /opt/keycloak/lib/quarkus/ /opt/keycloak/lib/quarkus/
  COPY --from=builder /opt/keycloak/lib/lib/main/quarkus-logging-gelf-2.7.0.Final.jar /opt/keycloak/lib/lib/main/quarkus-logging-gelf-2.7.0.Final.jar
  COPY --from=builder /opt/keycloak/providers/ /opt/keycloak/providers/
  
  WORKDIR /opt/keycloak
  ENV KEYCLOAK_ADMIN=admin
  ENV KEYCLOAK_ADMIN_PASSWORD=admin
  ENV KC_PROXY=edge
  
  ENTRYPOINT ["./bin/kc.sh", "start"]
  ```

We are sure the problem is not on the configuration of gelf/logstash since we use that service with the same parameters for other projects too without any issues. 

Do I have to activate or use some parameter in order to use the gelf jar ?

Thanks you and best regards

**Reply by sstepien** [reply]
Hi @tse00,

Have you managed to get integration with log-stash working.
We are migrating to Keycloak 17.0.0 and integration with log-stash is one of the outstanding issues. It is crucial for us as we run Keycloak cluster on k8s with ELK  stack.
 
Thanks for sharing and kind regards,

**Reply by tse00** [reply]
Hi @sstepien,

unfortunately not. Im using already Keycloak 17.0.1 (Quarkus 2.7.5) but I still have the issue Quarkus (on Keycloak) don't recognise the Gelf jar. Not sure if this is a quarkus or keycloak issue. I assume as soon as the jar is recognise everything else will work.
If you give it a try let me know if you come further or if you face the same problem.

Best regards

**Reply by errier** [reply]
@tse00 
Thanks, this helped. Maybe I can help you too.

About the jar files:
1. You also need to copy the `quarkus-logging-gelf-deployment-2.7.0.Final.jar` file to the providers directory for it to be picked up on runtime. Edit: You also may need `logstash-gelf-1.15.0.jar`.
2. There is indeed no need to also copy the same jar file to `/opt/keycloak/lib/lib/main`.

Two remarks about `KC_FEATURES=quarkus-logging-gelf`. 
1. The extension is called **logging-gelf**, not **quarkus-logging-gelf**. 
2. Besides the first point this environment variable is used for Keycloak features, if I understand the documentation correctly. So this won't work for non Keycloak features, I think. The documentation provides you with a list of extensions [it supports](https://www.keycloak.org/server/features#_relevant_options).

My logging now shows the extension:
```
[io.quarkus] (main) Installed features: [..., logging-gelf, ...]
```

**Reply by tse00** [reply]
@errier thanks a lot, is working now!!! Together with the quarkus-logging-gelf-deployment, I also needed to add the logstash-gelf. So to make this work 3 dependencies need to be copied to `/providers`

quarkus-logging-gelf-2.7.5.Final.jar
quarkus-logging-gelf-deployment-2.7.5.Final.jar
logstash-gelf-1.15.0.jar

**Reply by semangard** [reply]
It is possible to use ENV VARS into _quarkus.properties_ ?

**Reply by gunterze** [reply]
It's possible to use ENV VARS in `quarkus.properties`, e.g.  
```
quarkus.log.handler.gelf.enabled=${GELF_ENABLED:false}
```

**Reply by PJaroslaw** [reply]
Hi just wondering what I might have done wrong. I followed what's written in this post and added all the jars to /providers directory, but when I start up my keycloak-quarkus application, I don't see any logs being sent. Any ideas?

**Comment by gdelbos** [comment]
Hi here,

Using the information on this discussion, I was able to send Keycloak log to logstash using GELF. 
However, in the previous distrubution (using wildfly) we were able to configure a JUL filter to add some MDC informations.
This is really helpful as it will ensure that any Keycloak logs sent to logstash will be properly labelled as "keycloak" logs.

However, I cannot find a proper solution allowing me to do this.

Does anyone have a solution or a workaround ?

Thanks for your help.

**Reply by gdelbos** [reply]
Nevermind, I could use the "additional-field" configuration of quarkus gelf.
Is not exactly the same thing but it is an acceptable workaround.

eg. 

```
quarkus.log.handler.gelf.additional-field."mdc.service".value=${GELF_SERVICE_NAME:keycloak}
```
